File: cpp/0904M-水果成篮.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-10-14 16:39:50
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    int totalFruit(vector<int>& fruits) {
        int ans, k;
        ans = k = 0;
        unordered_map<int, int> basket;
        for (int i = 0; i < fruits.size(); i++) {
            basket[fruits[i]]++;
            while (basket.size() > 2) {
                basket[fruits[k]]--;
                if (basket[fruits[k]] == 0) basket.erase(fruits[k]);
                k++;
            }
            ans = max(ans, i - k + 1);
        }
        return ans;
    }
};

File: go/0904M-水果成篮.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-14 20:19:33
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func totalFruit(fruits []int) int {
	ans, k := 0, 0
	basket := make(map[int]int)
	for i := 0; i < len(fruits); i++ {
		basket[fruits[i]]++
		for len(basket) > 2 {
			basket[fruits[k]]--
			if basket[fruits[k]] == 0 {
				delete(basket, fruits[k])
			}
			k++
		}
		if ans < i-k+1 {
			ans = i - k + 1
		}
	}
	return ans
}
